### ISSUE TITLE
Data input issue.

### ISSUE THREAD

**Issue by Z0721Z** [primebo1/FAMNet#1]
Your excellent work has left me in awe. I have a question I'd like to ask:
  Is the `supp_imgs` not needed to be changed during the training process, and only the `supp_imgs` passed in the first round is used throughout the entire training process?

**Comment by primebo1** [comment]
> Your excellent work has left me in awe. I have a question I'd like to ask: Is the `supp_imgs` not needed to be changed during the training process, and only the `supp_imgs` passed in the first round is used throughout the entire training process?

Hi, thank you for your kind words.
To address your question: the support samples are not fixed throughout the training process. Instead, they are updated in each iteration to form a new episode. Both the support samples and the query images are randomly selected from 3D patient image cases, which are also randomly sampled in each iteration to ensure diversity during training.  Details can be found in `.\dataloaders\datasets.py`.

**Comment by Z0721Z** [comment]
You have a problem where the line `foreground_features = torch.stack(result_list)` in the `get_fg` function causes a mismatch in dimensions: one tensor has shape `[256, 115272]` and another has shape `[256, 114090]`. You think this issue might be caused by different data being selected due to varying masks. How can this issue be resolved?

**Comment by primebo1** [comment]
> [256, 115272

Hi, are you sure you are using our original code without any modifications? I have checked the `get_fg` function, and there are no errors when it is executed. Additionally, the `foreground_features` naturally differ for different samples, and when the model's `batch_size=1`, the `result_list` length is 1, so `torch.stack(result_list)` should not have any dimension alignment issues. If there are still problems, please provide detailed explanations or error messages so that I can reproduce the issue.

**Comment by Z0721Z** [comment]
> > [256, 115272
> 
> Hi, are you sure you are using our original code without any modifications? I have checked the `get_fg` function, and there are no errors when it is executed. Additionally, the `foreground_features` naturally differ for different samples, and when the model's `batch_size=1`, the `result_list` length is 1, so `torch.stack(result_list)` should not have any dimension alignment issues. If there are still problems, please provide detailed explanations or error messages so that I can reproduce the issue.

I replaced `tmp_mask[0]` with `tmp_mask.squeeze()`. This shouldn't be the cause of the issue, right?

**Comment by Z0721Z** [comment]
> > [256, 115272
> 
> Hi, are you sure you are using our original code without any modifications? I have checked the function, and there are no errors when it is executed. Additionally, the naturally differ for different samples, and when the model's , the length is 1, so should not have any dimension alignment issues. If there are still problems, please provide detailed explanations or error messages so that I can reproduce the issue.`get_fg``foreground_features``batch_size=1``result_list``torch.stack(result_list)`

    def get_fg(self, fts, mask):

        """
        :param fts: (1, C, H', W')
        :param mask: (1, H, W)
        :return:
        """

        mask = torch.round(mask)

        fts = F.interpolate(fts, size=mask.shape[-2:], mode='bilinear')

        if len(fts.shape) &lt; 4:
            fts = fts.unsqueeze(0)

        mask = mask.unsqueeze(1).bool()
        result_list = []

        for batch_id in range(fts.shape[0]):
            tmp_tensor = fts[batch_id]
            tmp_mask = mask[batch_id]
            
            print(tmp_tensor.shape)
            tmp_mask = tmp_mask.squeeze()

            print(tmp_mask.shape)
            foreground_features = tmp_tensor[:, tmp_mask]

            if foreground_features.shape[1] == 1:
                foreground_features = torch.cat((foreground_features, foreground_features), dim=1)

            result_list.append(foreground_features)

        foreground_features = torch.stack(result_list)

**Comment by Z0721Z** [comment]
> if len(fts.shape) &lt; 4:
>         fts = fts.unsqueeze(0)

if len(fts.shape) < 4:
        fts = fts.unsqueeze(0)

**Comment by primebo1** [comment]
> > > [256, 115272
> > 
> > 
> > Hi, are you sure you are using our original code without any modifications? I have checked the function, and there are no errors when it is executed. Additionally, the naturally differ for different samples, and when the model's , the length is 1, so should not have any dimension alignment issues. If there are still problems, please provide detailed explanations or error messages so that I can reproduce the issue.`` get_fg``foreground_features``batch_size=1``result_list``torch.stack(result_list) ``
> 
> ```
> def get_fg(self, fts, mask):
> 
>     """
>     :param fts: (1, C, H', W')
>     :param mask: (1, H, W)
>     :return:
>     """
> 
>     mask = torch.round(mask)
> 
>     fts = F.interpolate(fts, size=mask.shape[-2:], mode='bilinear')
> 
>     if len(fts.shape) &lt; 4:
>         fts = fts.unsqueeze(0)
> 
>     mask = mask.unsqueeze(1).bool()
>     result_list = []
> 
>     for batch_id in range(fts.shape[0]):
>         tmp_tensor = fts[batch_id]
>         tmp_mask = mask[batch_id]
>         
>         print(tmp_tensor.shape)
>         tmp_mask = tmp_mask.squeeze()
> 
>         print(tmp_mask.shape)
>         foreground_features = tmp_tensor[:, tmp_mask]
> 
>         if foreground_features.shape[1] == 1:
>             foreground_features = torch.cat((foreground_features, foreground_features), dim=1)
> 
>         result_list.append(foreground_features)
> 
>     foreground_features = torch.stack(result_list)
> ```



I replaced the `get_fg` function in my code with the one you provided, and it executed without any issues. I'm afraid I don't quite understand the specific problem you're encountering, but I suggest checking the input sizes at the entry of the function by using:  

```python
print(mask.shape)  
print(fts.shape)  
```  

The output should be:  
```
torch.Size([1, 256, 256])  
torch.Size([1, 512, 64, 64])  
```
